community/kernel/src/main/java/org/neo4j/kernel/impl/core/InMemoryTokenCache.java

```diff
@@ -43,11 +43,12 @@ public void clear()
         idToToken.clear();
     }
 
-    private static void putAndEnsureUnique( Map<String,Integer> nameToId, Token token, String tokenType )
+    private void putAndEnsureUnique( Map<String,Integer> nameToId, Token token, String tokenType )
     {
-        Integer previous;
-        if ( (previous = nameToId.put( token.name(), token.id() )) != null && previous != token.id() )
+        Integer previous = nameToId.putIfAbsent( token.name(), token.id() );
```

Comment:
The `putIfAbsent` of our `CopyOnWriteHashMap` is not thread-safe.

Actually, why don't we give `CopyOnWriteHashMap` an `alter` method that is synchronised and will take a lambda of `HashMap to void`. The `HashMap` given to the lambda is then the to-be-installed copy, and the lambda can perform any modification on it that it desires. This way we can easily make arbitrary changes to a `CopyOnWriteHashMap` in a thread-safe way. Then, if we relax the generics, we can use a single map to represent both `nameToId` and `idToToken`. This way we can easily update both mappings in a thread-safe way. And the two-maps-in-one would not conflict with each other, because they use different types for their keys.